### DISCUSSION TITLE
wrong payload does not raise 422 when all fields are optional

### DISCUSSION THREAD

**Discussion by MaelAbgrall** [pydantic/pydantic#2693] [Question] [answered]
Hello,

I randomly had an issue when doing unit test with pydantic: I send an invalid payload, but no error was raised.

the model looks like this:
```python
class PostUpdate(BaseModel):
  text: Optional[str]
  title: Optional[str]

  @validator('*')
  def validate_not_null(cls, value) -> str:
    """Ensure that if a field is set, it is not null."""
    if value is None:
      raise HTTPException(
        status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
        detail="Value should not be set to None. Either set a value, or do not declare it."
      )
    return value
```
The idea was that a user could update either the text, or the title, or both.

I mis-spelled a payload for my unit test (see below).
The odd thing is that this payload did not raise a 422 by pydantic, actually resulted in a query against the database. I found this because the tests showed the content was not updated.
```javascript
// wrong payload
{
  "test": "updated text"
}
//should have been 
{
  "text": "updated text"
}
```

Is there a way to prevent this?

**Comment by lesnik512** [comment, marked as answer]
Hello
If i'm not mistaking, you could use [validate-always](https://pydantic-docs.helpmanual.io/usage/validators/#validate-always)
Because now validator is not being triggered by validate_not_null if field is not set

Updated:
And for checking if some of the fields are set, you can use [root validator](https://pydantic-docs.helpmanual.io/usage/validators/#root-validators)

```python
class PostUpdate(BaseModel):
  text: Optional[str]
  title: Optional[str]

  @root_validator()
  def validate_not_null(cls, values) -> str:
    """Ensure that if a field is set, it is not null."""
    if values.get("text") is None and values.get("title") is None:
      raise HTTPException(
        status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
        detail="Value should not be set to None. Either set a value, or do not declare it."
      )
    return value
```

**Reply by MaelAbgrall** [reply]
Thanks for the links! The validate-always would defeat the purpose of the optional imho

It seems combining the two responses would be the best, to avoid extra attributes, and be sure that at least one field has been declared

cheers

**Comment by PrettyWood** [comment]
IMO you want to prevent extra attributes with
```py
class PostUpdate(BaseModel, extra=Extra.forbid):
```
or
```py
class PostUpdate(BaseModel):
    class Config:
        extra=Extra.forbid
```
Your issue here is that "test" field was allowed but simply ignored and you end up with the two default values at `None`.
You probably don't want to change your validators since you want to still allow the `None` default value

**Reply by MaelAbgrall** [reply]
I find this really odd to not be set to forbid by default? Is there an explanation behind this?

I'm considering creating an abstract class with this extra, and make all the other models inherit it

Thank you by the way =)

**Reply by PrettyWood** [reply]
Don't hesitate to mark this question as answered once you've tested and found the best solution for you

**Reply by MaelAbgrall** [reply]
I suppose there is a good reason for this extra parameter set to ignore by default, so I'll stay with the other one for now and keep this in mind if I need to do a more strict check.
Thank you!

**Reply by arogozhnikov** [reply]
@PrettyWood 
ok, so ... why forbid is not enforced by default? Can I somehow make this default (at least for me?)

it is not assigned, no warning raised, it just skipped ... I was kinda shocked because I deserialized previous model and was looking for problem in the wrong place for a day, while reason was - after renaming parameter was dropped.
